### PR TITLE
Check if nodes are biophysical before calling .morph

### PR DESCRIPTION
**Pull request by tomdele**

### CODE REVIEW

**Comment by tomdele** [comment]
removed most of the mocks from morph tests

**Comment by tomdele** [comment]
I did it the simple way. 
But I think it would be super nice in the future to be able to define subclasses of nodes with added Interfaces.

File: tests/test_morph.py [hunk 1]
```diff
@@ -32,18 +87,65 @@ def test_get_1(self):
...
+            decimal=6
+        )
+        assert len(actual) == 32
+        # swc file
+        # index       type         X            Y            Z       radius       parent
+        # 1           1    -0.320000     1.000000     0.000000     0.725000          -1
+        # 2           1    -0.320000     0.900000     0.000000     0.820000           1
+
+        # rotation around the x axis 90 degrees counter clockwise
+        expected = [
```
Comment (asanin-epfl) [inline tests/test_morph.py:113]: I'm puzzled here. We had a 90 degrees rotation of a point (101,102,103), yet we see that coordinates didn't change almost at all `[100.68, 102, 104.0, 0.725]`. Shouldn't at least one of coordinates become negative? What am I missing?

File: tests/test_morph.py [hunk 2]
```diff
@@ -32,18 +87,65 @@ def test_get_1(self):
...
+            [100.68, 102, 104.0, 0.725],
+            [100.68, 102, 103.9, 0.820]
+        ]
+        npt.assert_almost_equal(
+            expected, actual[:2]
+        )
+        npt.assert_almost_equal(expected, actual[:2])
+
+    def test_get_3(self):
```
Comment (asanin-epfl) [inline tests/test_morph.py:122]: just `test_get`?

File: tests/test_morph.py [hunk 3]
```diff
@@ -53,24 +155,26 @@ def test_get_2(self):
...
-    # should get cached object for 'morph-A'
-    assert test_obj.get(2) is morph1
-    # should get new object ('morph-B')
-    assert test_obj.get(3) is not morph1
-    # 'morph-A' was evicted from cache
-    assert test_obj.get(4) is not morph1
+    with patch.object(test_module.MorphHelper, '_is_biophysical', return_value=True):
+        test_obj = test_module.MorphHelper('morph-dir', nodes)
+        nm_load.side_effect = Mock
+        morph1 = test_obj.get(1)
```
Comment (asanin-epfl) [inline tests/test_morph.py:161]: I thought indexing in SONATA starts from 0? I understand it does not matter here as we use mocks but can we use indexing from 0, please?

File: tests/test_morph.py [hunk 4]
```diff
@@ -53,24 +155,26 @@ def test_get_2(self):
...
-    test_obj = test_module.MorphHelper('morph-dir', nodes)
-    nm_load.side_effect = Mock
-    morph1 = test_obj.get(1)
-    # same morphology, but no no caching
-    assert test_obj.get(2) is not morph1
\ No newline at end of file
+    with patch.object(test_module.MorphHelper, '_is_biophysical', return_value=True):
+        test_obj = test_module.MorphHelper('morph-dir', nodes)
+        nm_load.side_effect = Mock
+        morph1 = test_obj.get(1)
```
Comment (asanin-epfl) [inline tests/test_morph.py:178]: indexing from 0 here also please.

Comment (tomdele) [reply tests/test_morph.py:178]: good catch !

File: tests/test_morph.py (hunk 3 above)
Comment (tomdele) [reply tests/test_morph.py:161]: good catch !

File: tests/test_morph.py (hunk 2 above)
Comment (tomdele) [reply tests/test_morph.py:122]: change for `test_get_morphology_standard_rotation`

File: tests/test_morph.py (hunk 1 above)
Comment (tomdele) [reply tests/test_morph.py:113]: The rotation around the x axis is done before the translation. The translation being positive and far superior to the morph size you will only have positive number. 
Plus, the rotation basically keep X identical and switch the y and z axis for the morph. 
I added comments.

File: tests/test_morph.py [hunk 5]
```diff
@@ -32,18 +87,65 @@ def test_get_1(self):
         npt.assert_almost_equal(expected, actual[:2])
 
     def test_get_2(self):
```
[the remark this answers is not in the document]
Comment (tomdele) [reply tests/test_morph.py:89]: done. Changed to `test_get_morphology_simple_rotation`

File: tests/test_morph.py [hunk 6]
```diff
@@ -1,29 +1,84 @@
...
-        self.nodes.get.return_value = 'test'
-        actual = self.test_obj.get_filepath(42)
-        expected = str(self.morph_path / 'test.swc')
+        node_id = 0
+        assert self.nodes.get(node_id, properties="morphology") == "morph-A"
+        actual = self.test_obj.get_filepath(node_id)
+        expected = str(self.morph_path / 'morph-A.swc')
         assert actual == expected
 
     def test_get_1(self):
```
[the remark this answers is not in the document]
Comment (tomdele) [reply tests/test_morph.py:80]: sure makes sense. 
Change for : 
`test_get_morphology`

File: tests/test_morph.py [hunk 7]
```diff
@@ -32,18 +87,65 @@ def test_get_1(self):
...
+
+        # rotation around the x axis 90 degrees counter clockwise
+        expected = [
+            [100.68, 102, 104.0, 0.725],
+            [100.68, 102, 103.9, 0.820]
+        ]
+        npt.assert_almost_equal(
+            expected, actual[:2]
+        )
+        npt.assert_almost_equal(expected, actual[:2])
```
[the remark this answers is not in the document]
Comment (tomdele) [reply tests/test_morph.py:120]: bad cut and paste. Being removed.

File: tests/test_morph.py [hunk 8]
```diff
@@ -32,18 +87,65 @@ def test_get_1(self):
...
-        self.nodes.get.return_value = 'small_morph'
-        self.nodes.positions.return_value = pd.Series({
-            'x': 100.0,
-            'y': 200.0,
-            'z': 300.0,
-        })
-        self.nodes.orientations.return_value = -np.identity(3)
-        actual = self.test_obj.get(42, transform=True).points
+        node_id = 0
+        actual = self.test_obj.get(node_id, transform=True).points
```
[the remark this answers is not in the document]
Comment (tomdele) [reply tests/test_morph.py:91]: sure I will

File: tests/test_morph.py (hunk 4 above)
Comment (tomdele) [reply tests/test_morph.py]: with the mocking this is not a problem but yeah that was not consistent

File: tests/test_morph.py (hunk 1 above)
Comment (asanin-epfl) [reply tests/test_morph.py:114]: That clarifies things. Thank you.

**Pull request merged**